src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,609 @@
+            ["/usr/sbin/usermod", "-aG", MICROK8S_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/usermod", "-aG", LXD_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/chmod", "777", "/usr/local/bin"], check=True, timeout=30
+        )
+    except subprocess.SubprocessError as exc:
+        raise SystemUserConfigurationError from exc
+
+
+YQ_DOWNLOAD_URL_TMPL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/yq_linux_{BIN_ARCH}"
+)
+YQ_BINARY_CHECKSUM_URL = "https://github.com/mikefarah/yq/releases/latest/download/checksums"
+YQ_CHECKSUM_HASHES_ORDER_URL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/checksums_hashes_order"
+)
+YQ_EXTRACT_CHECKSUM_SCRIPT_URL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/extract-checksum.sh"
+)
+
+
+class ExternalPackageInstallError(Exception):
+    """Represents an error installilng external packages."""
+
+
+def _validate_checksum(file: Path, expected_checksum: str) -> bool:
+    """Validate the checksum of a given file.
+
+    Args:
+        file: The file to calculate checksum for.
+        expected_checksum: The expected file checksum.
+
+    Returns:
+        True if the checksums match. False otherwise.
+    """
+    sha256 = hashlib.sha256()
+    sha256.update(file.read_bytes())
+    return sha256.hexdigest() == expected_checksum
+
+
+BIN_ARCH_MAP: dict[Arch, str] = {Arch.ARM64: "arm64", Arch.X64: "amd64"}
+
+
+def _install_external_packages(arch: Arch) -> None:
+    """Install packages outside of apt.
+
+    Installs yarn, yq.
+
+    Args:
+        arch: The architecture to download binaries for. #TODO check bin arch
+
+    Raises:
+        ExternalPackageInstallError: If there was an error installing external package.
+    """
+    try:
+        subprocess.run(
+            ["/usr/bin/npm", "install", "--global", "yarn"], check=True, timeout=60 * 5
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/npm", "cache", "clean", "--force"], check=True, timeout=60
+        )  # nosec: B603
+        bin_arch = BIN_ARCH_MAP[arch]
+        yq_path_str = f"yq_linux_{bin_arch}"
+        # The URLs are trusted
+        urllib.request.urlretrieve(
+            YQ_DOWNLOAD_URL_TMPL.format(BIN_ARCH=bin_arch), yq_path_str
+        )  # nosec: B310
+        urllib.request.urlretrieve(YQ_BINARY_CHECKSUM_URL, "checksums")  # nosec: B310
+        urllib.request.urlretrieve(
+            YQ_CHECKSUM_HASHES_ORDER_URL, "checksums_hashes_order"
+        )  # nosec: B310
+        urllib.request.urlretrieve(
+            YQ_EXTRACT_CHECKSUM_SCRIPT_URL, "extract-checksum.sh"
+        )  # nosec: B310
+        # The output is <BIN_NAME> <CHECKSUM>
+        checksum = subprocess.check_output(  # nosec: B603
+            ["/usr/bin/bash", "extract-checksum.sh", "SHA-256", yq_path_str],
```

Comment:
Executing an externally downloaded shell script in pretty risky, we must avoid that

Reply:
Currently i'm not sure if there are ways to avoid download yq binary without having a less capable snapped yq (which should then be moved to cloud-init script since snaps cannot be installed within chroot).
Should we do so? (it may break other people's pipelines running yq since it will not be read from root dirs.)

Reply:
I don't think snap is a good alternative, as it's strictly confined and limit it's FS visibility. If there's no apt package, we might build is from source in the builder and copy the built binary ? Or create our own apt package for ya

Reply:
Changed to build YQ from source, thank you!